betty/generate/pool.py

```diff
@@ -0,0 +1,205 @@
+"""
+Provide the site generation multiprocessing pool.
+"""
+
+from __future__ import annotations
+
+import multiprocessing
+import os
+import queue
+from asyncio import run
+from concurrent import futures
+from concurrent.futures import Executor, ProcessPoolExecutor
+from contextlib import ExitStack
+
+from typing import (
+    Callable,
+    MutableSequence,
+    Self,
+    Concatenate,
+    Any,
+    TYPE_CHECKING,
+    ParamSpec,
+)
+
+import dill
+
+from betty.app import App
+from betty.asyncio import gather
+
+from betty.generate import GenerationContext
+from betty.project import Project
+
+if TYPE_CHECKING:
+    from types import TracebackType
+    import threading
+
+
+_GenerationProcessPoolTaskP = ParamSpec("_GenerationProcessPoolTaskP")
+
+_generation_process_pool_worker_setup: Callable[[], None] | None = None
+
+
+class _GenerationProcessPool:
+    """
+    Set up a worker process, before the worker starts performing tasks.
+
+    This may be used to modify the environment, set up mocks, etc.
+    """
+
+    def __init__(self, project: Project):
+        self._project = project
+        self._queue = multiprocessing.Manager().Queue()
+        self._cancel = multiprocessing.Manager().Event()
+        self._finish = multiprocessing.Manager().Event()
+        self._exit_stack = ExitStack()
+        self._executor: Executor | None = None
+        self._workers: MutableSequence[futures.Future[None]] = []
+        # @todo Ensure this is synchronized across processes.
+        self._count_total = 0
+
+    async def __aenter__(self) -> Self:
+        concurrency = os.cpu_count() or 2
+        # Avoid `fork` so as not to start worker processes with unneeded resources.
+        # Settle for `spawn` so all environments use the same start method.
+        executor = ProcessPoolExecutor(
+            max_workers=concurrency, mp_context=multiprocessing.get_context("spawn")
+        )
+        self._exit_stack.enter_context(executor)
+        # @todo Ensure we pass on the necessary dependencies
+        pickled_app_args = dill.dumps(
```

Comment:
Can we not make the App directly pickleable? It is only Project that has a tricky dependency in the form of App

Reply:
No, not if this would require Dill!